iwyu_include_picker.cc

```diff
@@ -1329,13 +1329,32 @@ vector<string> IncludePicker::GetCandidateHeadersForFilepathIncludedFrom(
   // ConvertToQuotedInclude because it avoids trouble when the same
   // file is accessible via different include search-paths, or is
   // accessed via a symlink.
-  const string& quoted_include_as_typed
-      = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
-  if (!quoted_include_as_typed.empty()) {
-    vector<string>::iterator it = std::find(retval.begin(), retval.end(),
-                                            quoted_includee);
-    if (it != retval.end())
-      *it = quoted_include_as_typed;
+  vector<string>::iterator it = std::find(retval.begin(), retval.end(),
+                                          quoted_includee);
+  if (it != retval.end()) {
+      const string& quoted_include_as_typed
+          = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
+      if (!quoted_include_as_typed.empty()) {
+          *it = quoted_include_as_typed;
+      } else {
+          string strippedIncludee = included_filepath;
+          // Directory of the file is implicitly in the search path.
+          // So if we have the full path, shorten it
+          const string implicitHeaderSearchPath =
+                  CanonicalizeHeaderSearchPath(GetParentPath(including_filepath));
+          if (StripLeft(&strippedIncludee, implicitHeaderSearchPath)) {
+              VERRS(6) << "Adding include in same directory for "
+                       << included_filepath << ": " << strippedIncludee
+                       << "\n";
+              // If the included path is absolute, we prefer the short one
+              // else we prefer the current one, which is probably relative to
+              // a search directory
+              if(IsAbsolutePath(StripQuotes(*it)))
+                  retval.insert(it, "\"" + strippedIncludee + "\"");
+              else
+                  retval.push_back("\"" + strippedIncludee + "\"");
+          }
+      }
```

Comment:
Yes, adding a slash indeed helps and tests pass. Running IWYU with the reverted code above on my real codebase shows however that I still get absolute paths. I've seen different behaviour when the cc file is passed as absolute vs relative. So I assume this is hat the test don't cover yet